lib/iris/common/metadata.py

```diff
@@ -0,0 +1,402 @@
+            emsg = "Require a subclass of {!r}, got {!r}."
+            raise TypeError(emsg.format(BaseMetadata.__name__, cls))
+
+        #: The metadata class to be manufactured by this factory.
+        self.cls = cls
+
+        # Initialise the metadata class fields in the instance.
+        for field in self.fields:
+            setattr(self, field, None)
+
+        # Populate with provided kwargs, which have already been verified
+        # by the factory.
+        for field, value in kwargs.items():
+            setattr(self, field, value)
+
+    def __eq__(self, other):
+        if not hasattr(other, "cls"):
+            return NotImplemented
+        match = self.cls is other.cls
+        if match:
+            match = self.values == other.values
+        return match
+
+    def __getstate__(self):
+        """Return the instance state to be pickled."""
+        return {field: getattr(self, field) for field in self.fields}
+
+    def __ne__(self, other):
+        match = self.__eq__(other)
+        if match is not NotImplemented:
+            match = not match
+        return match
+
+    def __reduce__(self):
+        """
+        Dynamically created classes at runtime cannot be pickled, due to not
+        being defined at the top level of a module. As a result, we require to
+        use the __reduce__ interface to allow 'pickle' to recreate this class
+        instance, and dump and load instance state successfully.
+
+        """
+        return (MetadataFactory, (self.cls,), self.__getstate__())
+
+    def __repr__(self):
+        args = ", ".join(
+            [
+                "{}={!r}".format(field, getattr(self, field))
+                for field in self.fields
+            ]
+        )
+        return "{}({})".format(self.__class__.__name__, args)
+
+    def __setstate__(self, state):
+        """Set the instance state when unpickling."""
+        for field, value in state.items():
+            setattr(self, field, value)
+
+    @property
+    def fields(self):
+        """Return the name of the metadata members."""
+        return self.cls._fields
+
+    @property
+    def values(self):
+        fields = {field: getattr(self, field) for field in self.fields}
+        return self.cls(**fields)
+
+    # Restrict factory to appropriate metadata classes only.
+    if not issubclass(cls, BaseMetadata):
+        emsg = "Require a subclass of {!r}, got {!r}."
+        raise TypeError(emsg.format(BaseMetadata.__name__, cls))
+
+    # Check whether kwargs have valid fields for the specified metadata.
+    if kwargs:
+        extra = [field for field in kwargs.keys() if field not in cls._fields]
+        if extra:
+            bad = ", ".join(map(lambda field: "{!r}".format(field), extra))
+            emsg = "Invalid {!r} field parameters, got {}."
+            raise ValueError(emsg.format(cls.__name__, bad))
+
+    # Define the name, (inheritance) bases and namespace of the dynamic class.
+    name = "Metadata"
```

Comment:
Could we call this `MetadataManager` instead?

And perhaps call this function `MetadataManagerFactory` instead of `MetadataFactory`?

Reply:
@lbdreyer Agreed. Let's do both for clarity.